lib/watchpool/watchpool.go

```diff
@@ -0,0 +1,169 @@
+	// leader times out and returns to it's caller. If there are "follower" Watch
+	// calls, one of them will receive the message and take over being the leader.
+	leaderDone chan struct{}
+	// refs is a reference count for how many outstanding Watch calls there are
+	// for this WatchSet. It is incremented and decremented atomically.
+	refs uint32
+}
+
+// Watch performs a shared watch on a memdb.WatchSet. It will deduplicate calls
+// with the same WatchSets (i.e. exactly the same nodes in the state store being
+// watched) and run only one ws.WatchCtx call regardless of the number of
+// concurrent callers.
+func (p *WatchPool) Watch(ctx context.Context, ws memdb.WatchSet) error {
+	key, err := p.wsKey(ws)
+	if err != nil {
+		// This shouldn't ever really happen but if we return immediately then it
+		// turns whatever bug causes it into a blocking query hot loop. Instead,
+		// revert to normal un-pooled blocking execution.
+		return ws.WatchCtx(ctx)
+	}
+
+	// Lazy initialize the map
+	if p.watchSets == nil {
+		p.watchSets = make(map[uint64]*wsEntry)
+	}
+
+	// Whatever happens, we need to come back and decrement the ref count and
+	// possibly clean up at the end.
+	defer p.watcherDone(key)
+
+	// See if the key already exists in the share map
+	p.Lock()
+	entry, ok := p.watchSets[key]
+	if !ok {
+		// Create a new entry
+		entry = &wsEntry{
+			ws:   ws,
+			done: make(chan struct{}),
+			// Buffer in leaderChan is important to not block leader if it timesout
+			// with no followers. We rely on it to not deadlock the initial call too.
+			leaderDone: make(chan struct{}, 1),
+			refs:       1,
+		}
+		p.watchSets[key] = entry
+
+		// Trigger one of the Watch goroutines for this set to become the initial
+		// leader. It will likely be us in the select below but it could be another
+		// concurrent call that races. Note that if the select below takes a
+		// non-leader case because our context is cancelled for example, then either
+		// another Watcher will become leader instead, or all watchers will return
+		// for another reason and the state will be cleaned up by watcherDone
+		// preserving invariant that there is always one leader Watch for a given
+		// entry in the map. This won't block because we buffer this chan.
+		entry.leaderDone <- struct{}{}
+	} else {
+		// Add ourselves to the ref count
+		entry.refs++
+	}
+	// Unlock now before processing any more
+	p.Unlock()
+
+	// We are a follower
+	select {
+	case <-ctx.Done():
+		// Our context is done, just return it's error.
+		return ctx.Err()
+
+	case <-entry.done:
+		// The leader notified of a change in the watchset.
+		return nil
+
+	case <-entry.leaderDone:
+		// The current leader timed out (or we just setup the state above) and we
+		// won the race to become next leader. We need to watch until our context is
+		// done
+		watchErr := entry.ws.WatchCtx(ctx)
+		if watchErr == context.Canceled || watchErr == context.DeadlineExceeded {
```

Comment:
I would explicitly check for unknown errors there... we only check for context.Canceled and context.DeadlineExceeded right?

Reply:
Right, any other error means that the WatchSet actually returned an error. In practice if you look in the current WatchSet code that can't happen, but if that code ever changes for example because WatchSet is empty and we decide to treat that as an error instead, then I think it would be correct to consider that error a fatal one and _not_ pass the leadership on to another goroutine which will also hit the same error etc.

Ideally here the semantics would be "is this an error that only affects _me_ or all watchers of this set". But I don't want to complicate it by trying to add special error types etc for now when there are in fact the only two errors we can ever see currently.

Would it help to comment that?